superscore/control_layers/core.py

```diff
@@ -0,0 +1,204 @@
+    Control Layer used to communicate with the control system, dispatching to
+    whichever shim is relevant.
+    """
+    shims: Dict[str, _BaseShim]
+
+    def __init__(self, *args, **kwargs):
+        self.shims = {
+            'ca': AiocaShim(),
+        }
+
+    def shim_from_pv(self, pv: str) -> _BaseShim:
+        """
+        Determine the correct shim to use for the provided ``pv``.
+        ``pv`` can optionally hold a protocol defining prefix such as "ca://" or
+        "pva://".  If no prefix is provided, will select the first available shim.
+
+        Parameters
+        ----------
+        pv : str
+            a PV address such as "MY:PREFIX:mtr1" or "pva://MY:PREFIX:dt"
+
+        Returns
+        -------
+        _BaseShim
+            The shim held by this ControlLayer for ``pv``'s protocol
+
+        Raises
+        ------
+        ValueError
+            If pv cannot be recognized or a matching shim cannot be found
+        """
+        split = pv.split("://", 1)
+        if len(split) > 1:
+            # We got something like pva://mydevice, so use specified comms mode
+            shim = self.shims.get(split[0], None)
+        else:
+            # No comms mode specified, use the default
+            shim = list(self.shims.values())[0]
+
+        if shim is None:
+            raise ValueError(f"PV is of an unsupported protocol: {pv}")
+
+        return shim
+
+    @singledispatchmethod
+    def get(self, pv: Union[str, list[str]]) -> Any:
+        """
+        Get the value(s) in ``pv``.
+        If a single pv is provided, will return a single value.
+        If a list of pvs is provided, will get the values for each asynchronously.
+
+        Parameters
+        ----------
+        pv : Union[str, list[str]]
+            The PV(s) to get values for.
+
+        Returns
+        -------
+        Any
+            The requested data
+        """
+        # Dispatches to _get_single and _get_list depending on type
+        print(f"PV is of an unsupported type: {type(pv)}. Provide either "
+              "a string or list of strings")
+
+    @get.register
+    def _get_single(self, pv: str) -> Any:
+        """Synchronously get a single ``pv``"""
+        return asyncio.run(self._get_one(pv))
+
+    @get.register
+    def _get_list(self, pv: list) -> Any:
+        """Synchronously get a list of ``pv``"""
+        async def gathered_coros():
+            coros = []
+            for p in pv:
+                coros.append(self._get_one(p))
+            return await asyncio.gather(*coros)
+
+        return asyncio.run(gathered_coros())
+
+    async def _get_one(self, pv: str):
+        """
+        Base async get function.  Use this to construct higher-level get methods
+        """
+        shim = self.shim_from_pv(pv)
+        return await shim.get(pv)
+
+    @singledispatchmethod
+    def put(
+        self,
+        pv: Union[str, list[str]],
+        value: Union[Any, list[Any]],
+        cb: Optional[Callable] = None
```

Comment:
```suggestion
        cb: Optional[Union[Callable, list[Callable]]] = None
```
As-is, the type hint doesn't match the pattern of the other parameters.  Not sure if that actually affects anything

Reply:
Nothing in the type-hints affects runtime behavior, it's all for static analysis and our convenience.  My analysis tools look directly at the dispatch methods for suggestions, and not the base dispatcher method.  

That being said I will definitely make this change 👍